src/tasks.ts

```diff
@@ -0,0 +1,331 @@
+
+    if (targetEpoch > slotToEpoch(clBlock.message.slot)) {
+        debug('Target epoch is not reached yet');
+        return;
+    }
+
+    // TODO: query report frame instead
+    const sourceEpoch = await Shared.ORACLE.lastConsolidatedEpoch({
+        blockTag,
+    });
+    debug('Source epoch is', sourceEpoch);
+
+    const artifact = readArtifact(Number(sourceEpoch), Number(targetEpoch));
+    if (!R.isNil(artifact)) {
+        debug('Report tree artifact already exists');
+        return;
+    }
+
+    const prevTreeCID = await Shared.ORACLE.treeCid({
+        blockTag,
+    });
+
+    if (R.isEmpty(prevTreeCID)) {
+        debug('No previous report tree CID found');
+    } else {
+        debug('Reading the previous tree from IPFS by CID', prevTreeCID);
+        const prevTree = loadTree(await catCID(prevTreeCID));
+
+        debug('Check the tree root');
+        const prevRoot = await Shared.ORACLE.reportRoot({
+            blockTag,
+        });
+
+        if (prevTree.root !== prevRoot) {
+            throw new Error('Tree root mismatch');
+        }
+
+        debug('Fill in the rewards cache');
+        for (const [_, [no, reward]] of prevTree.entries()) {
+            Cache.rewards.set(no, reward);
+        }
+    }
+
+    debug('Loading validators');
+    await loadValidators(Number(targetEpoch));
+
+    debug('Fetching duties');
+    // TODO: fix range
+    for (const epoch of R.range(Number(sourceEpoch), Number(targetEpoch))) {
+        await lookupCommittees(epoch);
+    }
+
+    debug('Check duties');
+    // TODO: fix range
+    for (const epoch of R.range(Number(sourceEpoch), Number(targetEpoch))) {
+        await checkEpochSlots(epoch);
+    }
+
+    // TODO: read the threshold from somewhere
+    Cache.excludeFromStats(belowThreshold(0.9));
+
+    const feeShares = await Shared.STETH.sharesOf(Shared.CONFIG.CSM_ADDRESS);
+    const distributed = distributeFees(feeShares + 42n);
+
+    const leafs = R.sortBy(R.prop(0), [...Cache.rewards.entries()]);
+    debug('Rewards distribution', leafs);
+    const tree = buildTree(leafs);
+    debug('Report tree root', tree.root);
+    saveArtifact(Number(sourceEpoch), Number(targetEpoch), tree);
+
+    // TODO: Simulate report sending
+    await Shared.ORACLE.connect(Shared.SIGNER).submitReport.staticCall(
+        targetEpoch,
+        tree.root,
+        distributed,
+        '', // CID TBA, but can be precalculated
+    );
+
+    const cid = await uploadTree(tree);
+    // TODO: send a report
+    await Shared.ORACLE.connect(Shared.SIGNER).submitReport(
+        targetEpoch,
+        tree.root,
+        distributed,
+        cid.toString(),
+    );
+
+    debug('Report sent, storing the artifact');
+    saveArtifact(Number(sourceEpoch), Number(targetEpoch), tree);
+}
+
+async function loadValidators(epoch: Epoch) {
+    const vals = await getModuleValidators(epochLastSlot(epoch));
+    for (const v of vals) {
+        if (v.validator.slashed) {
+            continue;
+        }
```

Comment:
It can lead to the case when validator was slashed in the middle of frame and we didn't count the half of the good working time

Reply:
Loses compared to earnings for the period makes the including of this validator is meaningless. But it's discussable.

Reply:
As discussed, let's keep slashed validators.